calico/felix/config.py

```diff
@@ -46,166 +46,311 @@
+        """
+        self.description = description
+        self.name = name
+        self.sources = sources
+        self.value = default
+        self.active_source = None
+        self.value_is_int = value_is_int
+
+    def set(self, value, source):
+        """
+        Set a value of a parameter - unless already set.
+        :param value: value
+        :param source: source; for example "Configuration file /etc/felix.cfg"
+        """
+        if self.active_source is None:
+            log.debug("Read value %r for %s (%s) from %r",
+                      value,
+                      self.name,
+                      self.description,
+                      source)
+
+            self.active_source = source
+
+            if self.value_is_int:
+                # Set value before the call to int, so the ConfigException has
+                # the right value if / when it goes wrong.
+                self.value = value
+                try:
+                    self.value = int(value)
+                except ValueError:
+                    raise ConfigException("Field was not integer",
+                                          self)
+            else:
+                # Calling str in principle can throw an exception, but it's
+                # hard to see how in practice, so don't catch and wrap.
+                self.value = str(value)
+        else:
+            log.warning("Ignore %r value for %s (%s) - already set from %r",
+                        source,
+                        self.name,
+                        self.description,
+                        self.active_source)
+
 
 class Config(object):
     def __init__(self, config_path):
         """
-        Create a config.
+        Create a config. This reads data from the following sources.
+        - Environment variables
+        - Configuration file
+        - per-host etcd (/calico/vX/config)
+        - global etcd (/calico/vX/host/<host>/config)
+
+        After object creation, the environment variables and config file have
+        been read, and the variables ETCD_ADDR and HOSTNAME have been set and
+        validated. The caller is then responsible for reading the remaining
+        config from etcd and calling report_etcd_config with the returned
+        values before the rest of the config structure can be used.
+
         :raises EtcdException
         """
-        self._KnownSections = set()
-        self._KnownObjects  = set()
+        self.parameters = {}
+
+        self.add_parameter("EtcdAddr", "Address and port for etcd",
+                           "localhost:4001", sources = [ENV, FILE])
+        self.add_parameter("FelixHostname", "Felix compute host hostname",
+                           socket.gethostname(), sources = [ENV, FILE])
+
+        self.add_parameter("StartupCleanupDelay", "Delay before cleanup starts",
+                           30, value_is_int=True)
+        self.add_parameter("MetadataAddr", "Metadata IP address or hostname",
```

Comment:
Happy for that to be handled in a separate issue rather than blocking this work from going in.